packages/fxa-auth-db-mysql/db-server/test/backend/db_tests.js

```diff
@@ -2722,6 +2722,53 @@ module.exports = function (config, DB) {
         )
       })
 
+      it('should cancel subscriptions', async () => {
+        await db.createAccountSubscription(account.uid, subscriptionIds[18], 'prod0', Date.now())
+        await db.createAccountSubscription(account.uid, subscriptionIds[19], 'prod1', Date.now())
+
+        const cancelledAt = Date.now()
+        await db.cancelAccountSubscription(account.uid, subscriptionIds[19], cancelledAt)
+
+        const subscriptions = await db.fetchAccountSubscriptions(account.uid)
+
+        assert.lengthOf(subscriptions, 2)
+        assert.deepEqual(
+          pickSet(subscriptions, 'subscriptionId'),
+          new Set([ subscriptionIds[18], subscriptionIds[19] ])
+        )
+        assert.deepEqual(
+          pickSet(subscriptions, 'cancelledAt'),
+          new Set([ null, cancelledAt ])
+        )
+      })
+
+      it('should fail to cancel a non-existent subscription', async () => {
+        let succeeded = false
+        try {
+          await db.cancelAccountSubscription(account.uid, subscriptionIds[20], Date.now())
+          succeeded = true
+        } catch (err) {
+          assert.equal(err.errno, 116)
+        }
+
+        assert.isFalse(succeeded)
+      })
+
+      it('should fail to cancel a cancelled subscription', async () => {
+        await db.createAccountSubscription(account.uid, subscriptionIds[21], 'prod0', Date.now())
+        await db.cancelAccountSubscription(account.uid, subscriptionIds[21], Date.now())
+
+        let succeeded = false
+        try {
+          await db.cancelAccountSubscription(account.uid, subscriptionIds[21], Date.now())
+          succeeded = true
```

Comment:
Could just [do an `assert.fail()`](https://github.com/mozilla/fxa/blob/bb029485ca1be82e51722e2874db32c6184c7d59/packages/fxa-auth-db-mysql/db-server/test/backend/db_tests.js#L2781) here, rather than tracking the `succeeded` flag?

Reply:
Yeah, happy to do that.

Fwiw, I like the clarity of a separate assertion for the succeeded case, because it gives clearer messaging if you're diagnosing test failures. The `assert.fail()` gets caught by the `catch` block, so the thing that actually fails the test is the subsequent `errno` assertion. That makes it kind of similar to what would happen if the db fails with the wrong error type, and I like my failures to be as unique as possible.

But I'm splitting hairs I realise, it's not misleading enough to actually cost anyone time.

Reply:
> The assert.fail() gets caught by the catch block, so the thing that actually fails the test is the subsequent errno assertion

Ah, I forgot about that. You already changed it, but now I also see the benefit of the way you had it 🤷‍♂